### DISCUSSION TITLE
Late binding of TypeVar?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5313] [Ideas]
I have a class (`MyWrapper`) that I want to make generic with respect to the return type of the method an input (`wrapped.forward`). I know that I could define a more verbose class (`MyVerboseWrapper`) where I have to manually provide `wrapped.forward` as an input... but this awkward and inconvenient given that `WrappedT` is bounded by `A`, which we already know will have a `forward()` method.

Is it infeasible to be able to perform a late binding of the `TypeVar` `ForwardR` inside the definition of `__init__` instead of having to do it in the signature of `__init__`?

```python
from __future__ import annotations

from typing import Any, Callable, Generic, TypeVar


class A:
    def forward(self) -> Any:
        ...


class B(A):
    def forward(self) -> int:
        return 0


class C(A):
    def forward(self) -> str:
        return ""


WrappedT = TypeVar("WrappedT", bound=A)
ForwardR = TypeVar("ForwardR")


class MyWrapper(Generic[WrappedT, ForwardR]):
    def __init__(self, wrapped: WrappedT) -> None:
        self.wrapped = wrapped

        # I want to bind ForwardR to the return type of wrapped.forward() here:
        wrapped_forward: Callable[..., ForwardR] = wrapped.forward

    def forward(self) -> ForwardR:
        return self.wrapped.forward()

a_wrapper = MyWrapper(B())
reveal_type(a_wrapper.forward())  # Type of "a_wrapper.forward()" is "Any"


class MyVerboseWrapper(Generic[WrappedT, ForwardR]):
    def __init__(self, wrapped: WrappedT, wrapped_foward: Callable[..., ForwardR]) -> None:
        self.wrapped = wrapped

    def forward(self) -> ForwardR:
        return self.wrapped.forward()


a_verbose_wrapper = MyVerboseWrapper(B(), B.forward)
reveal_type(a_verbose_wrapper.forward())  # Type of "a_verbose_wrapper.forward()" is "int"

```

**Comment by erictraut** [comment]
That's not how type variables work. There's no such thing as "late binding" of type variables in the Python static type system. (I'm not aware of the concept in any type system that supports generics.)

In Python, all type variables are solved within the context of a function or constructor call. Any unsolved type variables are bound to `Unknown`.

If I understand the problem you're trying to solve, there's a much easier and straightforward solution: a generic protocol.

```python
from typing import Generic, Protocol, TypeVar

T = TypeVar("T", covariant=True)
R = TypeVar("R")

class SupportsForward(Protocol[T]):
    def forward(self) -> T:
        ...

class B:
    def forward(self) -> int:
        return 0

class C:
    def forward(self) -> str:
        return ""

class MyWrapper(Generic[R]):
    def __init__(self, wrapped: SupportsForward[R]) -> None:
        self.wrapped = wrapped

    def forward(self) -> R:
        return self.wrapped.forward()

a_wrapper = MyWrapper(B())
reveal_type(a_wrapper.forward())  # int
```

**Reply by ringohoffman** [reply]
I like this solution, but I don't _just_ want to know that `wrapped` is a `SupportsForward[int]`, I'd still like to still know that it is a `B`. I guess that would be kind of like a [higher kinded type](https://github.com/python/typing/issues/548)? Like

```python

HKT = TypeVar("HKT", bound=SupportsForward)

class MyWrapper(Generic[HKT[R]]):
    def __init__(self, wrapped: HKT[R]) -> None:
        self.wrapped = wrapped

    def forward(self) -> R:
        return self.wrapped.forward()

a_wrapper = MyWrapper(B())
reveal_type(a_wrapper.wrapped)  # B
reveal_type(a_wrapper.forward())  # int
```

**Comment by povarhena** [comment]
good job
